biggraphite/drivers/_downsampling.py

```diff
@@ -19,9 +19,77 @@
 
 import array
 import math
+import Queue
+import threading
+
 
 _NaN = float("NaN")
 
+_MAX_WRITE_DELAY = 5 * 60  # in seconds
+
+
+class _Writer(threading.Thread):
+    def __init__(self, accessor):
+        super(_Writer, self).__init__()
+        self._accessor = accessor
+        self._queue = Queue.Queue()
+        self._must_stop = threading.Event()
+        self._avg_sleep = 0
+
+    def soon(self, metric, downsampled, on_done=None):
+        self._accessor.insert_downsampled_points_async(metric, downsampled, on_done)
+
+    def later(self, metric, downsampled, on_done=None):
+        if not self.__started.is_set():
+            self.start()
+        self._queue.put_nowait((metric, downsampled, on_done))
+
+    def stop(self):
+        self._must_stop.set()
+
+    def __del__(self):
+        self.stop()
+
+    def run(self):
+        """Insert aggregates periodically.
+
+        Write once in a while the aggregated points present in the dedicated queue.
+        The delay between two writes is automatically adjusted between 0s and
+        _MAX_WRITE_DELAY, wrt the frequency of the updates encountered.
+        There's a dedicated Event to stop this -otherwise infinite- loop.
+        """
+        max_requests = self._accessor.max_requests_per_connection
+
+        while True:
+            if self._must_stop.wait(timeout=self._avg_sleep):
+                break
+
+            per_metric = {}
+            while not self._queue.empty():
```

Comment:
You'll end up using a ton of memory by doing that. I suggest you store points in one or multiple dicts directly (to allow easy override)